File: 백준/스택/송예은/옥상 정원 꾸미기.py
N = int(input())
buildings = []
stack =[]
ans = 0

for i in range(N):
    building = int(input())
    buildings.append(building)

for i in range(len(buildings)-1,-1,-1):
    if len(stack) == 0:
        stack.append([buildings[i],i+1])
    else :
        while len(stack) > 0:
            if stack[-1][0] < buildings[i]:
                stack.pop()
            else :
                break

        stack.append([buildings[i],i+1])

        if len(stack) == 1:
            ans += len(buildings) -i-1
        else:
            ans += stack[-2][1] - i-1-1
print(ans)

File: 백준/스택/문종후/옥상정원꾸미기.js
let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n").map(Number);
let n = input.shift();
let Stack = [];
let answer = 0;
for (let i = 0; i < input.length; i++) {
  while (Stack.length) {
    if (Stack[Stack.length - 1] <= input[i]) {
      Stack.pop();
    } else break;
  }
  Stack.push(input[i]);

  answer += Stack.length - 1;
}

console.log(answer);
